server/src/main/java/io/deephaven/server/arrow/FlightServiceGrpcImpl.java

```diff
@@ -45,24 +55,102 @@ public class FlightServiceGrpcImpl extends FlightServiceGrpc.FlightServiceImplBa
         this.ticketRouter = ticketRouter;
         this.doExchangeFactory = doExchangeFactory;
+        this.basicAuthMarshaller = basicAuthMarshaller;
+        this.authRequestHandlers = authRequestHandlers;
     }
 
     @Override
     public StreamObserver<Flight.HandshakeRequest> handshake(
             StreamObserver<Flight.HandshakeResponse> responseObserver) {
-        return GrpcUtil.rpcWrapper(log, responseObserver, () -> {
-            throw GrpcUtil.statusRuntimeException(Code.UNIMPLEMENTED, "See deephaven-core#997; support flight auth.");
-        });
+        return GrpcUtil.rpcWrapper(log, responseObserver, () -> new HandshakeObserver(responseObserver));
+    }
+
+    private final class HandshakeObserver implements StreamObserver<Flight.HandshakeRequest> {
+
+        private final StreamObserver<Flight.HandshakeResponse> responseObserver;
+
+        private HandshakeObserver(StreamObserver<Flight.HandshakeResponse> responseObserver) {
+            this.responseObserver = responseObserver;
+        }
+
+        @Override
+        public void onNext(final Flight.HandshakeRequest value) {
+            // handle the scenario where authentication headers initialized a session
+            SessionState session = sessionService.getOptionalSession();
+            if (session != null) {
+                responseObserver.onCompleted();
+                return;
+            }
+
+            final AuthenticationRequestHandler.HandshakeResponseListener handshakeResponseListener =
+                    (protocol, response) -> {
+                        GrpcUtil.safelyExecute(() -> {
+                            responseObserver.onNext(Flight.HandshakeResponse.newBuilder()
+                                    .setProtocolVersion(protocol)
+                                    .setPayload(response)
+                                    .build());
+                        });
+                    };
+
+            final ByteString payload = value.getPayload();
+            final long protocolVersion = value.getProtocolVersion();
+            Optional<AuthContext> auth = Optional.empty();
+            try {
+                if (basicAuthMarshaller != null) {
+                    auth = basicAuthMarshaller.login(protocolVersion, payload, handshakeResponseListener);
+                }
+                if (auth.isEmpty()) {
+                    // TODO: anonymous auth via handshake?
+
+                    final WrappedAuthenticationRequest req = WrappedAuthenticationRequest.parseFrom(payload);
+                    final AuthenticationRequestHandler handler = authRequestHandlers.get(req.getType());
+                    if (handler != null) {
+                        auth = handler.login(protocolVersion, req.getPayload(), handshakeResponseListener);
+                    }
+                }
+            } catch (final AuthenticationException | InvalidProtocolBufferException e) {
+                auth = Optional.empty();
```

Comment:
this is one of the cases i could reasonably see sending more than just one "it didnt work" error - the contract for empty() from the providers is "no one knew how to handle your request"... eh i could go either way.

probably should log pb errors though?